Back up and restore windows metadata like created ts, file attribs like hidden, readonly, encrypted with a common extensible mechanism

aneesh-n opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
Back up and restore windows-specific metadata like file Created time and File Attributes like hidden, readonly, encrypted with a common extensible mechanism.

Restic did not back up multiple windows-specific meta-data.
Restic now backs up file Created time and File Attributes like hidden flag when backing up files and folders on windows.

Note: Separate PRs will be created for SecurityDescriptors, Windows Extended File Attributes, ADS.


Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Add support for CreationTime and File Attributes in Windows -
https://github.com/restic/restic/pull/3622
Directory and file attribute "hidden" not restored -
Closes https://github.com/restic/restic/issues/2075


Pull request 3622 attempts to use "CreationTime" and "FileAttributes" as keys in restic's extended attributes. 
However, windows can have its own extended attributes - https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_ea_information
Using these 2 values as keys would then cause conflicts while implementing Windows extended attributes, as there could be extended attributes themselves with those names.

The objective of this pull request is to handle numerous pending Windows specific features in a generic way.
This update adds a new "generic_attributes" map to the repository similar to the extended_attributes but is used exclusively for implementing custom OS specific features in restic. 
Originally, I had attempted to reuse the extended_attributes array, but that could cause conflicts with real extended_attributes on linux and windows as the custom attributes would be attempted to be restored as real extended_attributes. I also thought of keeping a new IsSpecial flag on extended_attributes which would have value "1" for custom attributes and "0" for real extended_attributes, in an attempt to keep changes to the repository to a minimum, however, that change would break if a new restic repository (with custom attributes added) would be attempted to be restored using an older restic executable.

Hence keeping a separate "generic_attributes" seemed to be the cleanest way to go and this update to the repository is fully backward compatible, i.e. existing repositories will continue to work with the new restic versions and the new features will start working on existing repositories with the updated restic version.
Updated repositories will also work with older restic executables (in which case the new attributes will simply be ignored).
Also, this update of having a "generic_attributes" key-value array could potentially be reused for other OS specific features in future.

I understand the concern with making any update to the repository format and that this requires a thorough review. Please let me know if there are any suggestions to improve this mechanism or a better way to implement this.

Checklist
---------
- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

aneesh-n commented on internal/restic/node.go:
```diff
@@ -26,6 +26,44 @@ type ExtendedAttribute struct {
+	// TypeFileAttribute is the GenericAttributeType used for storing file attributes for windows within the generic attributes map.
+	TypeFileAttribute GenericAttributeType = "WinFileAttrib"
+	// TypeCreationTime is the GenericAttributeType used for storing creation time within the generic attributes map.
+	TypeCreationTime GenericAttributeType = "WinCreationTime"
```
This 'TypeCreationTime' is Windows specific and needs a specific format.
BirthTime can be created as a separate GenericAttributeType for linux if and when needed.

MichaelEischer commented on internal/test/helpers.go:
```diff
@@ -47,10 +48,28 @@ func OKs(tb testing.TB, errs []error) {
+		} else if len(msgs) == 1 {
+			tb.Fatalf("\033[31m\n\n\t"+msgs[0]+"\n\n\texp: %#v\n\n\tgot: %#v\033[39m\n\n", exp, act)
+		} else {
+			args := make([]interface{}, len(msgs)-1)
```
you can replace the whole loop with `args := msgs[1:]`, and then inline it into `fmt.Sprintf`.

aneesh-n replied:
That actually results in - "cannot use msgs[1:] (value of type []string) as []any value in argument to fmt.Sprintf"
Not sure if the loop can be avoided.

However, I can still clean it up a bit - 

args := make([]interface{}, len(msgs)-1)
for i, msg := range msgs[1:] {
     args[i] = msg
}
msgString = fmt.Sprintf(msgs[0], args...)

aneesh-n replied:
Or I could create a separate Equalsf -

// Equals fails the test if exp is not equal to act.
func Equals(tb testing.TB, exp, act interface{}) {
	Equalsf(tb, exp, act, "")
}

// Equalsf fails the test if exp is not equal to act.
// msg is optional message with arguments to be printed.
func Equalsf(tb testing.TB, exp, act interface{}, msg string, args ...any) {
	tb.Helper()
	if !reflect.DeepEqual(exp, act) {
		tb.Fatalf("\033[31m\n\n\t"+fmt.Sprintf(msg, args...)+"\n\n\texp: %#v\n\n\tgot: %#v\033[39m\n\n", exp, act)
	}
}

MichaelEischer replied:
If think a single function would be more convenient.

aneesh-n replied:
Updated to single function

MichaelEischer commented on internal/test/helpers.go:
```diff
@@ -47,10 +48,28 @@ func OKs(tb testing.TB, errs []error) {
+		var msgString string
+		if len(msgs) == 0 {
+			tb.Fatalf("\033[31m\n\n\texp: %#v\n\n\tgot: %#v\033[39m\n\n", exp, act)
+		} else if len(msgs) == 1 {
```
is that special case actually necessary?

aneesh-n replied:
removed

MichaelEischer commented on internal/test/helpers.go:
```diff
@@ -47,10 +48,28 @@ func OKs(tb testing.TB, errs []error) {
+				args[i-1] = s
+			}
+			msgString = fmt.Sprintf(msgs[0], args...)
+			tb.Fatalf("\033[31m\n\n\t"+msgString+"\n\n\texp: %#v\n\n\tgot: %#v\033[39m\n\n", exp, act)
```
I'd like to avoid the duplicate `tb.Fatalf` calls. Please always use this format string, and just set msgString to an empty string if there are no messages.

aneesh-n replied:
Done

MichaelEischer commented on internal/test/helpers.go:
```diff
@@ -47,10 +48,28 @@ func OKs(tb testing.TB, errs []error) {
 // Equals fails the test if exp is not equal to act.
-func Equals(tb testing.TB, exp, act interface{}) {
+// msg is optional message to be printed, first param being format string and rest being arguments.
+func Equals(tb testing.TB, exp, act interface{}, msgs ...string) {
```
Good idea!

MichaelEischer reviewed (commented):
(This is just a preliminary question, not a full review in any way)

Is there a reason why you've copied(?) a large amount of code to `internal/fs/` instead of using github.com/Microsoft/go-winio as library? I didn't check how similar / different the fs code is compared to the library.

Using a separate set of attributes instead of trying to hack things into the extended attributes looks like a good idea. That's also reasonably forward-compatible such that older restic version can still restore a backup (minus the new attributes obviously).

I'd prefer to split this PR into smaller parts. At first glance extended attributes support would nicely fit into a PR and everything else into another one. What do you think?

aneesh-n commented:
The reason I clubbed ExtendedAttributes with this one is because to implement that for windows, it required some refactoring in node.go where we now use restoreExtendedAttributes and fillExtendedAttributes. And that change fits nicely with the addition of restoreGenericAttributes and fillGenericAttributes in all node_xx. If that seems acceptable that would be great, otherwise I could go ahead and split them.

aneesh-n commented:
W.r.t go-winio, its ea.go does not expose some of the functions we need like SetFileEA, GetFileEA, which we would have to create separately anyways. Similarly, sd.go does not expose SetFileSecurityDescriptor and GetFileSecurityDescriptor for a file path.
That package is used for backing up and restoring those attributes while creating a tar. 
The code here was only used go-winio as reference for the most part other than sections of ea.go. 
Very little code would be useable from that package directly even if it was imported and it felt unnecessarily bulky to use it, adding very little value in terms of maintainability.

MichaelEischer commented:
> The reason I clubbed ExtendedAttributes with this one is because to implement that for windows, it required some refactoring in node.go where we now use restoreExtendedAttributes and fillExtendedAttributes. And that change fits nicely with the addition of restoreGenericAttributes and fillGenericAttributes in all node_xx, making it easier to review together. If that seems acceptable that would be great, otherwise I could go ahead and split them.

Hmm, the PR in its current state is far too large to review it properly (I usually start complaining around 500+ lines of new code). Most changes for extended and generic attributes appear to be independent of each other, such that we don't gain much from keeping the in one PR. So please split the PRs. We could do the following: first split extended attributes into a separate PR and then once we're done rebase this PR on top of it.

> The code here only used go-winio as reference for the most part other than sections of ea.go.
> Very little code would be useable from that package directly even if it was imported and it felt unnecessarily bulky to use it, adding very little value in terms of maintainability.

In that case please update the copyright notices to give a rough idea which code is copied from go-winio and which isn't. As the notice was used as a file header, I've assumed that it covers the complete file and ended up confused after noticing that the corresponding files in go-winio are far shorter.

MichaelEischer commented:
Or we could start with the GenericAttributes CreationTime and FileAttributes, then add security descriptors in a follow up PR and finally add extended attributes.

aneesh-n commented:
Updated this PR. Will create the next ones, once this is done.

hgraeber commented:
I am working on a similar PR, but restricted to the security descriptor (owner, group, dacl and sacl) and extended the node with one member, a optional string named "sddl", which stores the SDDL of the file or folder.

My experience is, that the ``SeBackupPrivilege``, ``SeRestorePrivilege`` and ``SeSecurityPrivilege`` are not always required. The Owner of a file can backup and restore the security descriptor - except the SACL - without those Privileges and can give the permission to do that to others via the DACL, too.

Therefore I decided to not check for those Privileges, but let restic fail on backup ore restore, if privileges, ownership and /or or access rights are not sufficient. Additionally, I planned to have a cli option which can be used to specify which parts of the security descriptor shall be backuped or restored, defaulting to owner, group and DACL, so that restic can be used without hassle by ordinary users for their own files. Users do not care about SACLs anyway.

For backup and restore of files of many different users one has to be member of the groups "Administrators" or "Backup Operators". Those usually have the needed Privileges and can access the SACL, too.

For accessing the security descriptor I use the functions ``windows.GetNamedSecurityInfo`` and ``windows.SetNamedSecurityInfo``, which are path based instead of handle based.

aneesh-n commented:
> I am working on a similar PR, but restricted to the security descriptor (owner, group, dacl and sacl) and extended the node with one member, a optional string named "sddl", which stores the SDDL of the file or folder.
> 
> My experience is, that the `SeBackupPrivilege`, `SeRestorePrivilege` and `SeSecurityPrivilege` are not always required. The Owner of a file can backup and restore the security descriptor - except the SACL - without those Privileges and can give the permission to do that to others via the DACL, too.
> 
> Therefore I decided to not check for those Privileges, but let restic fail on backup ore restore, if privileges, ownership and /or or access rights are not sufficient. Additionally, I planned to have a cli option which can be used to specify which parts of the security descriptor shall be backuped or restored, defaulting to owner, group and DACL, so that restic can be used without hassle by ordinary users for their own files. Users do not care about SACLs anyway.
> 
> For backup and restore of files of many different users one has to be member of the groups "Administrators" or "Backup Operators". Those usually have the needed Privileges and can access the SACL, too.
> 
> For accessing the security descriptor I use the functions `windows.GetNamedSecurityInfo` and `windows.SetNamedSecurityInfo`, which are path based instead of handle based.

Thanks for the input.
The optional sddl/sd information can be added as a GenericAttribute once this PR is merged.

In Microsoft's go-winio package (check https://github.com/microsoft/go-winio/blob/bc421d9108ade8ae2f76c8ef3e90cd4ee690e2c0/backuptar/tar.go#L130), they have deprecating sddl-based usage of SecurityDescriptors and instead use base64 encoded raw descriptors. I had updated my implementation from sddl to raw accordingly.
"Maintaining old SDDL-based behavior for backward compatibility. All new tar headers written by this library will have raw binary for the security."

I considered the cli options as well but thought it might be adding unnecessary complications in the early implementation. But I do feel there may be valid scenarios where users may want to customize which parts of the SDs should be backed up or restored and if they should be skipped altogether during restore. The default however should be to grab all the SD information possible during backup and restore them.

Could you elaborate on why "Users do not care about SACLs anyway."? It seems important from a security, auditing and incident investigation perspective. I think there would be use cases where it is important.
"For backup and restore of files of many different users one has to be member of the groups "Administrators" or "Backup Operators". - Yes agreed, in which case the enable privilege calls would be successful.

Let me review and test out your remaining points regarding privileges and get back. 

I have already removed the SecurityDescriptor change from this PR, so this shouldn't affect the current PR.
Do you have a PR to help in better collaboration? I would be creating a PR for SecurityDescriptor implementation after this PR is merged.

hgraeber commented:
> Thanks for the input. The optional sddl/sd information can be added as a GenericAttribute once this PR is merged.
> 
> In Microsoft's go-winio package (check https://github.com/microsoft/go-winio/blob/bc421d9108ade8ae2f76c8ef3e90cd4ee690e2c0/backuptar/tar.go#L130), they have deprecating sddl-based usage of SecurityDescriptors and instead use base64 encoded raw descriptors. I had updated my implementation from sddl to raw accordingly. "Maintaining old SDDL-based behavior for backward compatibility. All new tar headers written by this library will have raw binary for the security."

I would prefer the SDDL aynway, because it some readable an better for testing.

> I considered the cli options as well but thought it might be adding unnecessary complications in the early implementation. But I do feel there may be valid scenarios where users may want to customize which parts of the SDs should be backed up or restored and if they should be skipped altogether during restore. Even a basic disable-sd-restore option may be a good first step. The default however should be to grab all the SD information possible during backup and restore them.
> 
> Could you elaborate on why "Users do not care about SACLs anyway."? It seems important from a security, auditing and incident investigation perspective.

It is not important, if a normal user want's to backup his own personal data and the SACL is the only part of the sd, which really needs special privileges.

>I think there would be use cases where it is important. "For backup and restore of files of many different users one has to be member of the groups "Administrators" or "Backup Operators". - Yes agreed, in which case the enable privilege calls would be successful.

Yes, it's different for the admins, they want backup the SACL, too. But those usually have the privileges to backup/restore the SACL.

> Let me review and test out your remaining points regarding privileges and get back.
> 
> I have already removed the SecurityDescriptor change from this PR, so this shouldn't affect the current PR. Do you have a PR to help in better collaboration? I would be creating a PR for SecurityDescriptor implementation after this PR is merged.

I think it's a good idea to make my PR at least halfway ready, so we can compare the solutions.

aneesh-n commented:
Agreed.
But regarding SDDL, there's a reason why they removed it -
"Remove use of SDDL in tar headers
Switching to using straight binary instead of SDDL in tar headers to avoid failures in converting domain SIDs into/out of SDDL."

While I agree SDDL is better for testing, the functionality is more important and you do not need to read SDs in normal usage anyways.
Will check if some test helpers can be created for interoperability with SDDLs.

aneesh-n commented:
Thanks for the suggestions. I did make the change to use GetNamedSecurityInfo and SetNameSecurityInfo which avoid the need for the file handle.
Also, I try to first do the backup/restore of the full SD with all needed privileges. If that fails, we fallback to a lower-level SD backup/restore.
In case of backup, without running as admin, we can backup owner, group, DACL.
However, I find that during restore, you can only restore the DACL. Even if you're trying to set the same owner and group, it fails, and you need to pass nil instead.
https://github.com/zmanda/zestic/blob/windows-sd-support/internal/fs/sd_windows.go

aneesh-n commented on internal/restic/node.go:
```diff
@@ -478,7 +573,33 @@ func (node Node) sameContent(other Node) bool {
 func (node Node) sameExtendedAttributes(other Node) bool {
-	if len(node.ExtendedAttributes) != len(other.ExtendedAttributes) {
+	// Converting the array of ExtendedAttributes to an array of Attribute
+	var attributes []Attribute
```
If we want to avoid creation of these temp []Attribute arrays we could copy the full method sameExtendedAttributes and create sameGenericAttributes. Maintainability vs performance hit. Which would be preferred here?

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+	return syscall.SetFileAttributes(pathPointer, attrs)
+}
+
+// fixEncryptionAttribute checks if a file needs to be marked encrypted and is not already encrypted, it sets
```
Fixed backup and restore of FILE_ATTRIBUTE_ENCRYPTED in windows.

aneesh-n replied:
Encrypted attribute can only be set when a file is being newly created. When calling `SetFileAttribute` with file attributes including encrypted attribute, windows simply ignores the encrypted (`FILE_ATTRIBUTE_ENCRYPTED`) attribute and applies the rest.
The other way to handle this would be to delete a file if it exists and recreate it with `FILE_ATTRIBUTE_ENCRYPTED` if needed in fileswriter, which was making fileswriter very complex and more so when adding ADS functionality to it.

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+//go:build windows
+// +build windows
+
```
Added comprehensive testcase for backup and restore of all combinations of readonly, hidden, system, archive, encrypted and testcase for overriding of the attributes during restore.
Made other fixes as necessary, like fixes for restore of readonly files and encryption flag addition/removal as needed to fix the failing test cases.
Apologies for the size of the test case code.

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -4,12 +4,20 @@
 package restorer
 
```
Added comprehensive testcase for backup and restore of all combinations of readonly, hidden, system, archive, encrypted and testcase for overriding of the attributes during restore.
Made other fixes as necessary, like fixes for restore of readonly files and encryption flag addition/removal as needed to fix the failing test cases.
Apologies for the size of the test case code.

MichaelEischer commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+	return res
+}
+
+func saveDirOrdered(t testing.TB, repo restic.Repository, namedNodes []NamedNode, inode uint64) restic.ID {
```
`saveDirOrdered`/`saveOrderedSnapshot` appears to massively duplicate code from `saveDir`. What has changed compared to that function? Why are those changes necessary? Why is it _not possible_ to extend `saveDir` as required?

aneesh-n replied:
This was a byproduct of backporting these test cases from a more detailed ADS test with FileAttribute combinations. In case of ADS, we need to enforce an order in which ads related nodes are restored in order to cover certain scenarios. That ordering is not needed here. 
I have now cleaned up the code and reused saveDir and saveSnapshot.

MichaelEischer reviewed (changes requested):
I'm partially through with the first review round. Given the size of the PR, I'm pretty sure that I've missed some important things. For `restorer_windows_test.go` I've stopped reading the code after the first half; it's just too much for one go. It also looks like quite a lot of code from `restorer_test.go` was copied and partially adapted, see the comment for details.

The commits will also have to be rebased and (partially) squashed (now or later on).

I've also started to wonder whether GenericAttributes should refer to a struct with fixed attributes or whether the current generic list of attributes is better. What's your take on that?

aneesh-n commented on internal/fs/file.go:
```diff
@@ -124,3 +124,18 @@ func RemoveIfExists(filename string) error {
+func ClearReadonly(path string) error {
+	// If file is not writable, make it writable
+	// Set the new file permissions
+	if err := os.Chmod(path, 0600); err != nil {
```
Renaming to ResetPermissions. This seemed to be cleanest OS agnostic way of making the file writable again.

aneesh-n commented on internal/restic/node.go:
```diff
@@ -20,12 +20,76 @@ import (
 
+// GenericAttribute is a tuple storing the name and value pairs used internally by restic to provide support for OS-specific functionalities.
+// eg. For windows this is used for CreationTime, File Attributes like hidden etc.
+type GenericAttribute struct {
```
Using a common struct call Attribute now. It does clean up the code but makes the types intermixable, which I guess is a fairly harmless compromise.

aneesh-n commented on internal/restic/node.go:
```diff
@@ -20,12 +20,76 @@ import (
+)
+
+// When you create new GenericAttributeTypes for any OS, add an entry in this map.
+var genericAttributesForOS = map[string][]OSType{
```
Done

aneesh-n commented on internal/restic/node.go:
```diff
@@ -665,3 +810,43 @@ func (node *Node) fillTimes(stat *statT) {
+		} else {
+			//If genericAttributesForOS in node.go does not know about this attribute, then the repository may have been created by a newer version which has a newer GenericAttributeType.
+			msg := "WARNING: Found an unrecognized generic attribute in the repository: %s. You may need to upgrade to latest version of restic.%s"
+			_, err := fmt.Fprintf(os.Stderr, msg, genericAttributeName, "\n")
```
Agreed. 
For now, I can keep it only in the debug log. Any suggestions for how this can be informed to the user more explicitly?

aneesh-n commented on internal/restic/node.go:
```diff
@@ -665,3 +810,43 @@ func (node *Node) fillTimes(stat *statT) {
+// if called concurrently.
+func checkGenericAttributeNameNotHandledAndPut(value string) bool {
+	// Check if the key exists
+	if _, ok := unknownGenericAttributesHandlingHistory[value]; ok {
```
Fixed

aneesh-n commented on internal/restic/node_aix.go:
```diff
@@ -34,3 +37,16 @@ func Listxattr(path string) ([]string, error) {
+
+// restoreGenericAttributes is no-op on AIX.
+func (node *Node) restoreGenericAttributes(_ string) error {
+	for _, attr := range node.GenericAttributes {
```
Done

aneesh-n commented on internal/restic/node.go:
```diff
@@ -665,3 +810,43 @@ func (node *Node) fillTimes(stat *statT) {
+
+// checkGenericAttributeNameNotHandledAndPut checks if the GenericAttributeType name entry
+// already exists and puts it in the map if not. Not syncing this code as it is not a
+// critical operation. At most it will print a few extra warnings in the beginning
```
Fixed

aneesh-n replied:
Fixed

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+func (node Node) restoreGenericAttributes(path string) (err error) {
+	for _, attr := range node.GenericAttributes {
+		if errGen := attr.restoreGenericAttribute(path); errGen != nil {
+			err = fmt.Errorf("Error restoring generic attribute for: %s : %v", path, errGen)
```
Combining all errors into single error now. Did not want one error to prevent us from attempting to restore remaining generic attributes.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+		//Add Creation Time
+		node.appendGenericAttribute(getCreationTime(fi, path))
+	}
+	return true, err
```
This was left out because I removed the SecurityDescriptor handling code. This err will eventually get used then.
For now returning nil.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+
+// getFileAttributes gets the value for the GenericAttribute TypeFileAttribute
+func getFileAttributes(fileattr uint32) (fileAttribute GenericAttribute) {
+	fileAttrData := UInt32ToBytes(fileattr)
```
I thought about this but decided it's better to store all attributes during backup. Windows ignores the bits that it cannot restore and applies the rest. 
For example, encrypted attribute can only be set when a file is being newly created. When calling SetFileAttribute with encrypted attribute, windows simply ignores the encrypted (FILE_ATTRIBUTE_ENCRYPTED) attribute and applies the rest.
So, if we were to add the special handling for encrypted attributes in future i.e. with the encryptFile call, we could retrospectively allow previously backed up attributes to work during restores.
Hence, it's better to keep the entire set of attributes for the files during backup.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+// getFileAttributes gets the value for the GenericAttribute TypeFileAttribute
+func getFileAttributes(fileattr uint32) (fileAttribute GenericAttribute) {
+	fileAttrData := UInt32ToBytes(fileattr)
+	fileAttribute = NewGenericAttribute(TypeFileAttribute, fileAttrData)
```
done

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+
+// restoreGenericAttribute restores the generic attributes for windows like File Attributes,
+// Created time etc.
+func (attr GenericAttribute) restoreGenericAttribute(path string) error {
```
Now that we are using a common Attribute struct, this does not apply. Using a function instead - `func restoreGenericAttribute(attr Attribute, path string) error `.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+				// If existing file already has readonly or system flag, encrypt file call fails.
+				// We have already cleared readonly flag, clearing system flag if needed.
+				// The readonly and system flags will be set again at the end of this func if they are needed.
+				err = fs.ClearSystem(path)
```
We could, but that would add an additional system call for each and every file being restored. Assuming restore of system files is a relatively rare scenario compared to normal files, this felt like the greater good. Makes sense?

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+			err = decryptFile(pathPointer)
+			if err != nil {
+				if fs.IsAccessDenied(err) {
+					// If existing file already has readonly or system flag, encrypt file call fails.
```
Fixed

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+// marked encrypted, it removes the FILE_ATTRIBUTE_ENCRYPTED.
+func fixEncryptionAttribute(path string, attrs uint32, pathPointer *uint16) (err error) {
+	if attrs&windows.FILE_ATTRIBUTE_ENCRYPTED != 0 {
+		err = encryptFile(pathPointer)
```
Correct

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+
+// handleCreationTime gets the creation time from the data and sets it to the file/folder at
+// the specified path.
+func handleCreationTime(path string, data []byte) (err error) {
```
done

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +107,208 @@ func (s statT) ctim() syscall.Timespec {
+	}()
+
+	var inputData bytes.Buffer
+	inputData.Write(data)
```
Removed

aneesh-n commented on internal/restorer/fileswriter.go:
```diff
@@ -50,19 +50,34 @@ func (w *filesWriter) writeToFile(path string, blob []byte, offset int64, create
+					}
+					f, err = openFileWithTruncWrite(path)
+				} else {
+					return nil, err
```
Removed

aneesh-n commented on internal/restorer/fileswriter.go:
```diff
@@ -50,19 +50,34 @@ func (w *filesWriter) writeToFile(path string, blob []byte, offset int64, create
 		if createSize >= 0 {
-			flags = os.O_CREATE | os.O_TRUNC | os.O_WRONLY
+			// First check if file already exists
+			f, err = openFileWithTruncWrite(path)
```
Removed the extra step.

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+	tempDir := t.TempDir()
+	//Using the temp dir creation time as the test creation time for the test file and folder
+	creationTime := getCreationTime(t, tempDir)
+	runGenericAttributesTest(tempDir, t, restic.TypeCreationTime, creationTime)
```
Done

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+		//system
+		restic.UInt32ToBytes(syscall.FILE_ATTRIBUTE_SYSTEM),
+		//archive
+		restic.UInt32ToBytes(syscall.FILE_ATTRIBUTE_ARCHIVE),
```
Done

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+	}
+	runGenericAttributesTestForNodes(expectedNodes, tempDir, t, genericAttributeName, genericAttributeExpected)
+}
+func runGenericAttributesTestForNodes(expectedNodes []restic.Node, tempDir string, t *testing.T, genericAttr restic.GenericAttributeType, genericAttributeExpected []byte) {
```
Done

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+	}
+}
+
+func prepareGenericAttributesTestForNodes(tempDir string, testNode restic.Node, t *testing.T, genericAttr restic.GenericAttributeType) (string, []byte) {
```
Done

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+	if testNode.Type == "file" {
+
+		testFile, err := os.Create(testPath)
+		if err != nil {
```
Done

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+	if err != nil {
+		t.Fatal(errors.Wrapf(err, "Could not get NodeFromFileInfo for path: %s", testPath))
+	}
+	genericAttrThroughNodeFromFileInfo := nodeFromFileInfo.GetGenericAttribute(genericAttr)
```
Done

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,193 @@
+		t.Fatal(errors.Wrapf(err, "Could not Lstat for path: %s", path))
+	}
+
+	attrib, success := fi.Sys().(*syscall.Win32FileAttributeData)
```
Done

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+	IsDirectory bool
+}
+
+func generatePermutations(n int, prefix []bool) [][]bool {
```
Done

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+		dirInfo := NodeInfo{
+			DataStreamInfo: DataStreamInfo{
+				name: fileName,
+				data: "Main file data stream.",
```
No removed this. 
These test cases were back ported from a more detailed ADS test case and ADS streams on directories can have data. This should be removed though.

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+
+		//Run test
+		t.Run(testName, func(t *testing.T) {
+			mainFilePath := runAttributeTests(t, dirInfo, dirInfo.attributes)
```
There are many nested calls and there isn't a way to make the flow very convenient for all paths.
Anyways I've attempted to reordered them in order of their first usage.

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+
+			//Check directory exists
+			_, err1 := os.Stat(mainFilePath)
+			rtest.Assert(t, !errors.Is(err1, os.ErrNotExist), "The file "+dirInfo.name+" does not exist")
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+	return err
+}
+
+func SetupWithAttributes(t *testing.T, nodeInfo NodeInfo, testDir string, existingFileAttr Attributes) (*Restorer, []int) {
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+	t.Helper()
+	if nodeInfo.Exists {
+		if !nodeInfo.IsDirectory {
+			os.MkdirAll(path.Join(testDir, nodeInfo.parentDir), os.ModeDir)
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+		}
+	}
+
+	if nodeInfo.Exists {
```
Done

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +41,621 @@ func getBlockCount(t *testing.T, filename string) int64 {
+		mode = os.FileMode(2147484159)
+	} else {
+		if attributes != nil && attributes.ReadOnly {
+			mode = os.FileMode(292)
```
Done

aneesh-n commented:
> I'm partially through with the first review round. Given the size of the PR, I'm pretty sure that I've missed some important things. For `restorer_windows_test.go` I've stopped reading the code after the first half; it's just too much for one go. It also looks like quite a lot of code from `restorer_test.go` was copied and partially adapted, see the comment for details.
> 
> The commits will also have to be rebased and (partially) squashed (now or later on).
> 
> I've also started to wonder whether GenericAttributes should refer to a struct with fixed attributes or whether the current generic list of attributes is better. What's your take on that?

I've finished resolving the review comments.
I think the current approach with generic list of arguments could be more flexible to accommodate future use cases for other OSs as well. Also, after using a common Attribute struct the code seems cleaner now.

Let me know how we should rebase/squash. Should we squash all commits in this PR?

MichaelEischer commented:
Sorry for being slow to respond, we had to make a bugfix release of restic, so your PR had to wait. I'll try to find time for it in the next few days.

MichaelEischer reviewed (changes requested):
Thanks for all the changes, I'm mostly happy with the code, there are just a few nitpicks in the comments below. The main point that still has to be discussed is the JSON format of the GenericAttributes, see below.

> I think the current approach with generic list of arguments could be more flexible to accommodate future use cases for other OSs as well. Also, after using a common Attribute struct the code seems cleaner now.

While thinking about the pros and cons of either variant, I've arrived at the following requirements:

The list of possible attribute names is predetermined by restic. The user cannot influence the name of the attributes. However, later restic versions may add new attribute names. An older restic version should be able to recognize that there are some attributes which it does not support. This should also allow the `rewrite` command process such Trees, at least as long as no changes to the generic attributes are necessary. The order of attributes is actually irrelevant, it should be up to the implementation to select a suitable order of processing the attributes.

 I've noticed that there are actually quite a few possible designs here:

1. `GenericAttributes []Attribute` satisfies most requirements, although the attribute order is not necessary and attributes have to be decoded manually (which introduces the risk of missing format checks, and results in leaking encoding details into the tests like the usage of `UInt32ToBytes` in the restorer test.)
2. `GenericAttributes map[string][]byte` is very similar, but gets rid of the unnecessary attribute ordering. The JSON library sorts map keys prior to serialization, such that the resulting JSON encoding is deterministic.
3. `GenericAttributes map[string]json.RawMessage` would allow using the JSON decoder to decode values instead of having to do it manually. It still requires code that accesses the generic attribute to know how to deserialize it. This could be combined with a function `genericAttributesToWindowsAttrs(map[string]json.RawMessage) (WindowsAttributes, []string)` that decodes the Windows attributes and returns a list of unknown attributes.
4. `GenericAttributes struct { Field1 Type1, Field2 Type2}`. Is a much better fit for the property that all attribute values are determined by restic. However, due to limitations of the Go JSON decoder, it would drop unknown fields.
5. `GenericAttributesRaw json.RawMessage` + `genericAttributes struct { Field1 Type1, Field2 Type2} json:"-"`. The Go JSON decoder would fill in `GenericAttributesRaw`, which is then further decoded into genericAttributes by `UnmarshalJSON`. `GenericAttributesRaw` would only be overwritten if the generic attributes are modified. Detecting unknown attributes would be possible, but based on a quick test it would be extremely messy to implement.

Interestingly, the last three approaches can generate identical JSON representations, variant 2 is also quite similar. Variant 1 has a rather distinct format.

Based on these variants, I'm currently slightly preferring variant 3 (`map[string]json.RawMessage`). Although that is still open for debate.

> Let me know how we should rebase/squash. Should we squash all commits in this PR?

The variant with the least effort would be to split the PR into about three parts:
- helper function extensions (errors package and the test helper; one separate commit for each of them)
- the GenericAttributes code (one large commit is probably OK; adding the GenericAttributes mechanism first and adding the Windows code later would also be an option. Although I don't think it's really worth the effort)
- the restorer changes & test

aneesh-n commented on internal/restic/node.go:
```diff
@@ -20,12 +20,45 @@ import (
+	// TypeCreationTime is the GenericAttributeType used for storing creation time within the generic attributes map.
+	TypeCreationTime GenericAttributeType = "WinCreationTime"
+
+	//Generic Attributes for other OS types should be defined here.
```
Fixed

aneesh-n commented on internal/restic/node.go:
```diff
@@ -225,6 +270,24 @@ func (node Node) restoreMetadata(path string) error {
+		}
+	}
+
+	//Moving RestoreTimestamps and restoreExtendedAttributes calls above as for readonly files
```
Added space after `//`.
Regarding the ordering for readonly which only applies to windows, I guess it's fine to change the order in the common code (as currently implemented) as it would not affect the functionality for other OSs. Let me know if you disagree.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +106,201 @@ func (s statT) ctim() syscall.Timespec {
+func restoreGenericAttribute(attr Attribute, path string) error {
+	switch attr.Name {
+	case string(TypeFileAttribute):
+		return handleFileAttributes(path, attr.Value)
```
Fixed

aneesh-n commented on internal/restic/node_windows_test.go:
```diff
@@ -0,0 +1,179 @@
+	path := t.TempDir()
+	fi, err := os.Lstat(path)
+	test.OK(t, errors.Wrapf(err, "Could not Lstat for path: %s", path))
+	creationTimeAttribute := restic.GetCreationTime(fi, path)
```
Fixed. Had to change package for node_test as well as parseTime was being used from there.

aneesh-n commented on internal/restorer/fileswriter.go:
```diff
@@ -50,19 +50,28 @@ func (w *filesWriter) writeToFile(path string, blob []byte, offset int64, create
+			}
 		} else {
-			flags = os.O_WRONLY
+			flags := os.O_WRONLY
```
Fixed

aneesh-n commented on internal/restorer/fileswriter.go:
```diff
@@ -50,19 +50,28 @@ func (w *filesWriter) writeToFile(path string, blob []byte, offset int64, create
+		var err error
 		if createSize >= 0 {
-			flags = os.O_CREATE | os.O_TRUNC | os.O_WRONLY
+			f, err = openFileWithCreate(path)
```
Fixed

aneesh-n commented on internal/restorer/restorer.go:
```diff
@@ -204,8 +204,19 @@ func (res *Restorer) restoreHardlinkAt(node *restic.Node, target, path, location
+		if err != nil {
+			return err
+		}
+		wr, err = openFileWithTruncWrite(target)
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +40,516 @@ func getBlockCount(t *testing.T, filename string) int64 {
+	}
+
+	streams := []DataStreamInfo{}
+	if !nodeInfo.IsDirectory {
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +40,516 @@ func getBlockCount(t *testing.T, filename string) int64 {
+func createEncryptedFileWriteData(filepath string, fileInfo NodeInfo) error {
+	var ptr *uint16
+	ptr, err := windows.UTF16PtrFromString(filepath)
+	if err == nil {
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +40,516 @@ func getBlockCount(t *testing.T, filename string) int64 {
+			fileattr |= windows.FILE_ATTRIBUTE_DIRECTORY
+		}
+
+		fileAttrData := restic.UInt32ToBytes(fileattr)
```
Fixed

aneesh-n commented on internal/restorer/restorer_windows_test.go:
```diff
@@ -33,3 +40,516 @@ func getBlockCount(t *testing.T, filename string) int64 {
+			IsDirectory: true,
+		}
+
+		existingDirAttribute := []Attributes{}
```
Fixed

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +106,201 @@ func (s statT) ctim() syscall.Timespec {
+	}()
+
+	var creationTime syscall.Filetime
+	creationTime.LowDateTime = binary.LittleEndian.Uint32(data[0:4])
```
Fixed

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -83,3 +106,201 @@ func (s statT) ctim() syscall.Timespec {
+// handleFileAttributes gets the File Attributes from the data and sets them to the file/folder
+// at the specified path.
+func handleFileAttributes(path string, data []byte) (err error) {
+	attrs := binary.LittleEndian.Uint32(data)
```
Fixed

aneesh-n commented:
> Based on these variants, I'm currently slightly preferring variant 3 (`map[string]json.RawMessage`). Although that is still open for debate.

I've tried to refactor the `GenericAttributes` part along variant 3 using Convention Over Configuration and reflection to make it easier to extend and maintain. Let me know what you think.

Conventions -
OS Specific Attribute types should follow the convention `<OS>Attributes`.
`GenericAttributeType` should follow the convention `<OS specific attribute type>.<attribute name>`
The attributes in OS specific attribute types must be pointers as we want to distinguish nil values and not create `GenericAttributes` for them.
When new `GenericAttributeType` are defined, they must be added in the init function as well.

All `GenericAttributeType` are still defined in Node because we want to be able to distinguish between attributes belonging to different OS and unknown (newer restic version) attributes.

I've also fixed all other review comments.

MichaelEischer commented on internal/restorer/fileswriter.go:
```diff
@@ -127,12 +128,10 @@ func (w *filesWriter) writeToFile(path string, blob []byte, offset int64, create
 
 // openFileWithCreate opens the file with os.O_CREATE flag along with os.O_TRUNC and os.O_WRONLY.
 func openFileWithCreate(path string) (file *os.File, err error) {
```
Please also fully inline `openFileWithCreate` and `openFileWithTruncWrite` into the calling functions. To understand `writeToFile` it is necessary to know the exact implementation of both functions, so it's easier to understand if the `os.OpenFile` calls are directly included in `writeToFile` and are not part of a helper function.

aneesh-n replied:
Done

MichaelEischer reviewed (changes requested):
I'm somewhat undecided regarding whether it's worth the added complexity to use reflection for handling the generic attributes. as I don't have a strong opinion in either direction, I'd be fine with keeping it.  However, what I'd like to change is the naming of the GenericAttributeTypes. These are currently written in CamelCase, which looks very out of place, when taking a look at the JSON representation. All other fields of a `Node` are written in snake_case, so I'd like to stay consistent in that regard. That would convert `WindowsAttributes.CreationTime` to `windows.creation_time` (the attributes part is redundant). So for `GenericAttributesToOSAttrs` let's just pass in `windows` as key prefix and use `field.Tag.Get("generic")` to retrieve the snake_case name (see internal/options/options.go, the struct fields then have to be annotated using `generic:"creation_time" ).

Alternatively, we could also just drop the reflection code and manually implement the conversion code for the two attributes. Just go with the variant, that you think is less work to implement, if it becomes necessary we can still change the implementation in the future.

Besides that I only have a few nitpicks left, see below. Thanks for keeping up with the flood of review comments so far!

aneesh-n commented on internal/restic/node.go:
```diff
@@ -754,28 +778,34 @@ func (node *Node) fillTimes(stat *statT) {
-			debug.Log("Ignoring a generic attribute found in the repository: %s which may not be compatible with your OS. Compatible OS: %v", genericAttributeName, value)
+			// If genericAttributesForOS contains an entry but we still got here, it means the specific node_xx.go for the current OS did not handle it and the repository may have been originally created on a different OS.
+			// The fact that node.go knows about the attribute, means it is not a new attribute. This may be a common situation if a repo is used across OSs.
+			debug.Log("Ignoring a generic attribute found in the repository: %s which may not be compatible with your OS. Compatible OS: %s", genericAttributeType, os)
```
Yes, default behavior is to ignore attributes from a different OS, unless special handling is added for those attributes when possible, which is not the case for any of the attributes yet.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -108,16 +117,31 @@ func (s statT) ctim() syscall.Timespec {
+		return fmt.Errorf("error parsing generic attribute for: %s : %v", path, err)
+	}
+	for name := range node.GenericAttributes {
+		if errGen := restoreGenericAttribute(name, windowsAttributes, path, warn); errGen != nil {
```
Fixed

aneesh-n commented on internal/restic/node.go:
```diff
@@ -790,3 +820,67 @@ func checkGenericAttributeNameNotHandledAndPut(value GenericAttributeType) bool
 }
+
+// GenericAttributesToOSAttrs gets the os specific attribute from the generic attribute using reflection
+func GenericAttributesToOSAttrs(attrs map[GenericAttributeType]json.RawMessage, attributeType reflect.Type, attributeValuePtr *reflect.Value) (unknownAttribs []GenericAttributeType, err error) {
```
It doesn't have to be, but it was causing lint errors like -
func `genericAttributesToOSAttrs` is unused (unused)
func `osAttrsToGenericAttributes` is unused (unused) etc.
because these are helper methods meant to be used by the specific node_xx, they are unused from within node.

I now added // nolint:unused on these functions instead. I didn't want to move them to node_windows because they are meant to be reused for other OSs in future.

aneesh-n commented on internal/restic/node.go:
```diff
@@ -603,6 +588,45 @@ func sameAttributes(attributes []Attribute, otherAttributes []Attribute) bool {
+		}
+
+		// Check if the JSON.RawMessage values are equal byte by byte
+		if !bytesEqual(value1, value2) {
```
Fixed

aneesh-n commented on internal/restic/node.go:
```diff
@@ -20,43 +22,52 @@ import (
-	// TypeCreationTime is the GenericAttributeType used for storing creation time within the generic attributes map.
-	TypeCreationTime GenericAttributeType = "WinCreationTime"
+	// TypeCreationTime is the GenericAttributeType used for storing creation time for windows files within the generic attributes map.
+	TypeCreationTime GenericAttributeType = "WindowsAttributes.CreationTime"
```
Fixed

MichaelEischer reviewed (commented):
I have one last nitpick, see ~~below~~ (edit: apparently above). Then please squash the commits into a few commits as suggested earlier.

aneesh-n commented on internal/restic/node_windows.go:
```diff
@@ -126,76 +127,47 @@ func (node Node) restoreGenericAttributes(path string, warn func(msg string)) (e
 func genericAttributesToWindowsAttrs(attrs map[GenericAttributeType]json.RawMessage) (windowsAttributes WindowsAttributes, unknownAttribs []GenericAttributeType, err error) {
 	waValue := reflect.ValueOf(&windowsAttributes).Elem()
-	unknownAttribs, err = GenericAttributesToOSAttrs(attrs, reflect.TypeOf(windowsAttributes), &waValue)
+	unknownAttribs, err = genericAttributesToOSAttrs(attrs, reflect.TypeOf(windowsAttributes), &waValue, runtime.GOOS)
```
Done

aneesh-n committed:
Add CombineErrors helper function

aneesh-n committed:
Add optional messages for Equals helper

aneesh-n committed:
Support windows metadata using generic attribs

Add new generic_attributes attribute in Node.
Use the generic attributes to add support for creation time and file attributes like hidden, readonly, encrypted in windows. Handle permission errors for readonly files in windows.
Handle backup and restore of encrypted attributes using windows system calls.

aneesh-n committed:
Add ability to report warnings to terminal

Report warnings to terminal when unrecognized generic attributes are found in the repository.

aneesh-n committed:
Handle readonly empty files in windows

aneesh-n committed:
Add tests for generic attribute changes

aneesh-n committed:
Add changelog and update docs for windows attr

aneesh-n commented:
> I have one last nitpick, see ~below~ (edit: apparently above). Then please squash the commits into a few commits as suggested earlier.

I'm done with the changes and rebasing and squashing the commits. I'll create the Security Descriptor PR next followed by Extended Attribs for Windows and finally ADS.

Appreciate the detailed review and all the suggestions to improve the code. Thank you!

MichaelEischer reviewed (approved):
LGTM. Thanks a lot for all the work!

One last thing: please try to keep the follow up PRs smaller in size (aim for about 500 lines). I usually try to review the whole PR at once, which at the current size requires a few hours of uninterrupted time which can be rather hard to find.
